# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

In [1]:
import pandas as pd
mkt = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv')

In [2]:
mkt

Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0               0  DK49336     Arizona              4809.216960       No   
1               1  KX64629  California              2228.525238       No   
2               2  LZ68649  Washington             14947.917300       No   
3               3  XL78013      Oregon             22332.439460      Yes   
4               4  QA50777      Oregon              9025.067525       No   
...           ...      ...         ...                      ...      ...   
10905       10905  FE99816      Nevada             15563.369440       No   
10906       10906  KX53892      Oregon              5259.444853       No   
10907       10907  TL39050     Arizona             23893.304100       No   
10908       10908  WA60547  California             11971.977650       No   
10909       10909  IV32877         NaN              6857.519928      NaN   

       Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0         Basic   College           2/18/11         Employed      M  ...   
1         Basic   College           1/18/11       Unemployed      F  ...   
2         Basic  Bachelor           2/10/11         Employed      M  ...   
3      Extended   College           1/11/11         Employed      M  ...   
4       Premium  Bachelor           1/17/11    Medical Leave      F  ...   
...         ...       ...               ...              ...    ...  ...   
10905   Premium  Bachelor           1/19/11       Unemployed      F  ...   
10906     Basic   College            1/6/11         Employed      F  ...   
10907  Extended  Bachelor            2/6/11         Employed      F  ...   
10908   Premium   College           2/13/11         Employed      F  ...   
10909     Basic  Bachelor            1/8/11       Unemployed      M  ...   

       Number of Open Complaints Number of Policies     Policy Type  \
0                            0.0                  9  Corporate Auto   
1                            0.0                  1   Personal Auto   
2                            0.0                  2   Personal Auto   
3                            0.0                  2  Corporate Auto   
4                            NaN                  7   Personal Auto   
...                          ...                ...             ...   
10905                        NaN                  7   Personal Auto   
10906                        0.0                  6   Personal Auto   
10907                        0.0                  2  Corporate Auto   
10908                        4.0                  6   Personal Auto   
10909                        0.0                  3   Personal Auto   

             Policy  Renew Offer Type  Sales Channel  Total Claim Amount  \
0      Corporate L3            Offer3          Agent          292.800000   
1       Personal L3            Offer4    Call Center          744.924331   
2       Personal L3            Offer3    Call Center          480.000000   
3      Corporate L3            Offer2         Branch          484.013411   
4       Personal L2            Offer1         Branch          707.925645   
...             ...               ...            ...                 ...   
10905   Personal L1            Offer3            Web         1214.400000   
10906   Personal L3            Offer2         Branch          273.018929   
10907  Corporate L3            Offer1            Web          381.306996   
10908   Personal L1            Offer1         Branch          618.288849   
10909   Personal L1            Offer4            Web         1021.719397   

       Vehicle Class Vehicle Size Vehicle Type  
0      Four-Door Car      Medsize          NaN  
1      Four-Door Car      Medsize          NaN  
2                SUV      Medsize            A  
3      Four-Door Car      Medsize            A  
4      Four-Door Car      Medsize          NaN  
...              ...          ...          ...  
10905     Luxury Car      Medsize            A  
10906  Four-Door Car   

In [3]:
mkt.columns = mkt.columns.str.lower().str.replace(' ', '_')

In [4]:
mkt = mkt.drop(columns=['unnamed:_0'])

In [5]:
mkt.isnull().sum()

customer                            0
state                             631
customer_lifetime_value             0
response                          631
coverage                            0
education                           0
effective_to_date                   0
employmentstatus                    0
gender                              0
income                              0
location_code                       0
marital_status                      0
monthly_premium_auto                0
months_since_last_claim           633
months_since_policy_inception       0
number_of_open_complaints         633
number_of_policies                  0
policy_type                         0
policy                              0
renew_offer_type                    0
sales_channel                       0
total_claim_amount                  0
vehicle_class                     622
vehicle_size                      622
vehicle_type                     5482
dtype: int64

In [6]:
mkt = mkt.drop_duplicates(keep='first')

In [7]:
mkt[mkt.duplicated(subset='customer')]

customer       state  customer_lifetime_value response  coverage  \
334    AV76907         NaN              4282.947971      NaN     Basic   
441    FV22585      Oregon              6960.920371       No  Extended   
554    AU23837      Oregon              5010.379310       No     Basic   
598    OW33876  Washington              2476.480866       No     Basic   
644    SK12432     Arizona              2388.353882       No     Basic   
...        ...         ...                      ...      ...       ...   
10890  VQ78876      Oregon              4660.399573       No     Basic   
10894  EX63393         NaN              6314.492009      NaN  Extended   
10899  WU60905  California              8610.667489       No     Basic   
10905  FE99816      Nevada             15563.369440       No   Premium   
10909  IV32877         NaN              6857.519928      NaN     Basic   

                  education effective_to_date employmentstatus gender  income  \
334                 College           2/26/11         Employed      F   40864   
441                 College            2/4/11         Employed      M   74487   
554                Bachelor           1/19/11         Employed      F   64014   
598                 College           2/14/11         Employed      F   47918   
644                Bachelor           2/14/11       Unemployed      M       0   
...                     ...               ...              ...    ...     ...   
10890              Bachelor            2/3/11         Employed      F   89834   
10894                Master           2/27/11         Employed      F   66947   
10899  High School or Below            2/6/11       Unemployed      F       0   
10905              Bachelor           1/19/11       Unemployed      F       0   
10909              Bachelor            1/8/11       Unemployed      M       0   

       ... number_of_open_complaints number_of_policies     policy_type  \
334    ...                       1.0                  1   Personal Auto   
441    ...                       0.0                  3   Personal Auto   
554    ...                       0.0                  7  Corporate Auto   
598    ...                       NaN                  1  Corporate Auto   
644    ...                       0.0                  1   Personal Auto   
...    ...                       ...                ...             ...   
10890  ...                       0.0                  1  Corporate Auto   
10894  ...                       0.0                  6   Personal Auto   
10899  ...                       0.0                  9  Corporate Auto   
10905  ...                       NaN                  7   Personal Auto   
10909  ...                       0.0                  3   Personal Auto   

             policy  renew_offer_type  sales_channel  total_claim_amount  \
334     Personal L3            Offer1          Agent          166.937747   
441     Personal L1            Offer1    Call Center           16.711329   
554    Corporate L3            Offer1          Agent          295.871784   
598    Corporate L3            Offer2    Call Center          246.256648   
644     Personal L1            Offer2            Web          677.970934   
...             ...               ...            ...                 ...   
10890  Corporate L3            Offer1         Branch           43.155950   
10894   Personal L1            Offer3          Agent          374.400000   
10899  Corporate L1            Offer3            Web          532.800000   
10905   Personal L1            Offer3            Web         1214.400000   
10909   Personal L1            Offer4            Web         1021.719397   

       vehicle_class vehicle_size vehicle_type  
334              SUV      Medsize            A  
441    Four-Door Car      Medsize          NaN  
554              NaN          NaN            A  
598     Two-Door Car      Medsize            A  
644              NaN          NaN            A  
...              ...          ...   

In [8]:
pd.set_option('display.max_columns', None)

mkt_duplicates = mkt[mkt.duplicated(subset='customer', keep=False)]

mkt_duplicates_sorted = mkt_duplicates.sort_values(by='customer')

mkt_duplicates_sorted.head(30)

customer       state  customer_lifetime_value response  coverage  \
10376  AA10041  California              7901.744238       No     Basic   
3960   AA10041         NaN              7901.744238      NaN     Basic   
8731   AA10041  California              7901.744238       No     Basic   
432    AA11235      Nevada              2568.843396       No     Basic   
6044   AA11235      Nevada              2568.843396       No     Basic   
5078   AA56476     Arizona              5595.389905      Yes     Basic   
10422  AA56476     Arizona              5595.389905      Yes     Basic   
5435   AA71604     Arizona             11986.592120       No  Extended   
4601   AA71604     Arizona             11986.592120       No  Extended   
6360   AB21519  California              3861.486269      Yes  Extended   
1781   AB21519         NaN              3861.486269      NaN  Extended   
3119   AB26022      Oregon              7998.347343       No     Basic   
3212   AB26022      Oregon              7998.347343       No     Basic   
430    AB96670         NaN              2393.915369      NaN     Basic   
2834   AB96670  California              2393.915369      Yes     Basic   
10534  AD28349     Arizona              4637.544647       No     Basic   
2501   AD28349     Arizona              4637.544647       No     Basic   
6792   AE74442  California              7379.456823       No     Basic   
7527   AE74442  California              7379.456823       No     Basic   
8833   AE74442  California              7379.456823       No     Basic   
8987   AF47139      Oregon              8406.385414       No     Basic   
7094   AF47139      Oregon              8406.385414       No     Basic   
8007   AF47139         NaN              8406.385414      NaN     Basic   
3487   AH41449      Nevada              7907.133288       No     Basic   
2535   AH41449         NaN              7907.133288      NaN     Basic   
8719   AI49712  Washington              2593.897887       No     Basic   
802    AI49712  Washington              2593.897887       No     Basic   
5402   AI49712  Washington              2593.897887       No     Basic   
5906   AI58313      Nevada             36946.063930       No     Basic   
8410   AI58313      Nevada             36946.063930       No     Basic   

                  education effective_to_date employmentstatus gender  income  \
10376  High School or Below            2/2/11       Unemployed      M       0   
3960   High School or Below            2/2/11       Unemployed      M       0   
8731   High School or Below            2/2/11       Unemployed      M       0   
432                Bachelor           1/28/11    Medical Leave      F   11167   
6044               Bachelor           1/28/11    Medical Leave      F   11167   
5078   High School or Below            2/5/11         Employed      F   74454   
10422  High School or Below            2/5/11         Employed      F   74454   
5435                 Master           1/17/11         Employed      F   87560   
4601                 Master           1/17/11         Employed      F   87560   
6360                College            1/6/11         Employed      F   93272   
1781                College            1/6/11         Employed      F   93272   
3119   High School or Below           1/10/11          Retired      M   26487   
3212   High School or Below           1/10/11          Retired      M   26487   
430                 College           1/31/11       Unemployed      M       0   
2834                College           1/31/11       Unemployed      M       0   
10534  High School or Below           2/23/11       Unemployed      M       0   
2501   High School or Below           2/23/11       Unemployed      M       0   
6792               Bachelor           1/21/11         Employed      F   61584   
7527               Bachelor           1/21/11         Employed      F   61584   
8833               Bachelor           1/21/11         Employed      F   61584   
8987      

In [9]:
mkt_duplicates_sorted.shape

(3255, 25)

In [10]:
mkt = mkt.groupby('customer').apply(lambda x: x.fillna(x.mode().iloc[0] if not x.mode().empty else x)).reset_index(drop=True)

In [11]:
mkt.isnull().sum()

customer                            0
state                               0
customer_lifetime_value             0
response                            0
coverage                            0
education                           0
effective_to_date                   0
employmentstatus                    0
gender                              0
income                              0
location_code                       0
marital_status                      0
monthly_premium_auto                0
months_since_last_claim             0
months_since_policy_inception       0
number_of_open_complaints           0
number_of_policies                  0
policy_type                         0
policy                              0
renew_offer_type                    0
sales_channel                       0
total_claim_amount                  0
vehicle_class                       0
vehicle_size                        0
vehicle_type                     4531
dtype: int64

In [13]:
mkt['vehicle_type'].unique()

array(['A', nan], dtype=object)

In [14]:
mkt = mkt.drop(columns=['vehicle_type'])

In [15]:
mkt = mkt.drop_duplicates(keep='first')

In [16]:
mkt[mkt.duplicated(keep=False)]

Empty DataFrame
Columns: [customer, state, customer_lifetime_value, response, coverage, education, effective_to_date, employmentstatus, gender, income, location_code, marital_status, monthly_premium_auto, months_since_last_claim, months_since_policy_inception, number_of_open_complaints, number_of_policies, policy_type, policy, renew_offer_type, sales_channel, total_claim_amount, vehicle_class, vehicle_size]
Index: []

In [17]:
mkt.duplicated().sum()

0

In [18]:
mkt[mkt.duplicated(subset='customer')]

Empty DataFrame
Columns: [customer, state, customer_lifetime_value, response, coverage, education, effective_to_date, employmentstatus, gender, income, location_code, marital_status, monthly_premium_auto, months_since_last_claim, months_since_policy_inception, number_of_open_complaints, number_of_policies, policy_type, policy, renew_offer_type, sales_channel, total_claim_amount, vehicle_class, vehicle_size]
Index: []

In [19]:
mkt.columns

Index(['customer', 'state', 'customer_lifetime_value', 'response', 'coverage',
       'education', 'effective_to_date', 'employmentstatus', 'gender',
       'income', 'location_code', 'marital_status', 'monthly_premium_auto',
       'months_since_last_claim', 'months_since_policy_inception',
       'number_of_open_complaints', 'number_of_policies', 'policy_type',
       'policy', 'renew_offer_type', 'sales_channel', 'total_claim_amount',
       'vehicle_class', 'vehicle_size'],
      dtype='object')

In [42]:
mkt['customer_lifetime_value'].unique()

array([ 7901.744238,  2568.843396, 24127.50402 , ...,  3196.578455,
        3256.766388, 20714.94043 ])

In [20]:
mkt['gender'].unique()

array(['M', 'F'], dtype=object)

In [21]:
mkt['state'].unique()

array(['California', 'Nevada', 'Washington', 'Arizona', 'Oregon'],
      dtype=object)

In [22]:
mkt['response'].unique()

array(['No', 'Yes'], dtype=object)

In [23]:
mkt['coverage'].unique()

array(['Basic', 'Premium', 'Extended'], dtype=object)

In [24]:
mkt['education'].unique()

array(['High School or Below', 'Bachelor', 'College', 'Master', 'Doctor'],
      dtype=object)

In [25]:
mkt['effective_to_date'].unique()

array(['2/2/11', '1/28/11', '1/26/11', '1/15/11', '2/11/11', '1/11/11',
       '2/5/11', '1/17/11', '1/30/11', '1/4/11', '1/6/11', '1/10/11',
       '2/12/11', '2/4/11', '2/18/11', '2/7/11', '1/31/11', '1/9/11',
       '2/3/11', '1/19/11', '2/1/11', '1/3/11', '2/21/11', '2/20/11',
       '1/1/11', '2/23/11', '2/22/11', '1/8/11', '1/14/11', '1/20/11',
       '1/21/11', '2/10/11', '1/18/11', '2/17/11', '2/28/11', '1/23/11',
       '1/27/11', '1/16/11', '1/5/11', '2/26/11', '1/29/11', '2/19/11',
       '2/8/11', '1/24/11', '2/25/11', '2/14/11', '2/13/11', '2/24/11',
       '2/9/11', '2/27/11', '1/2/11', '2/16/11', '1/7/11', '1/12/11',
       '1/25/11', '1/13/11', '2/6/11', '2/15/11', '1/22/11'], dtype=object)

In [26]:
mkt['employmentstatus'].unique()

array(['Unemployed', 'Medical Leave', 'Employed', 'Retired', 'Disabled'],
      dtype=object)

In [27]:
mkt['location_code'].unique()

array(['Suburban', 'Urban', 'Rural'], dtype=object)

In [28]:
mkt['marital_status'].unique()

array(['Single', 'Married', 'Divorced'], dtype=object)

In [29]:
mkt['monthly_premium_auto'].unique()

array([113,  73,  71, 105, 240, 107, 103,  98, 177, 110,  95,  81, 112,
       131,  64, 129,  70, 117,  96,  86,  67, 125,  94,  65, 115, 102,
        74, 228,  68,  88,  92, 109, 229,  62,  66,  63,  72,  61, 104,
        97,  90, 185, 100, 108,  69,  87,  79, 101, 122, 121, 106, 114,
       116,  99,  84,  93, 283,  85,  83, 242, 138, 195, 183,  91, 119,
        76, 192,  80, 118,  89, 135,  78, 173, 184, 130, 133, 193,  77,
       128, 189, 111, 255, 126, 238, 123, 191, 148, 143, 154,  82, 284,
       181, 144, 124, 156, 226, 234, 202, 180, 132, 137, 127, 161, 201,
       239, 170, 134, 171, 159, 136, 139, 285, 153, 182, 190, 151, 235,
       199, 218, 197, 186, 244, 219, 198, 271, 223, 150, 187, 196, 208,
       225, 253, 168, 158, 296, 152, 217, 146, 210, 287, 232, 247, 256,
       245, 274, 231, 142, 194, 215, 188, 268, 213, 214, 149, 145, 206,
       281, 172, 207, 167, 160, 165, 155, 275, 298, 220, 276, 249, 212,
       204, 141, 164, 176, 230, 211, 140, 222, 169, 179, 174, 16

In [30]:
mkt['number_of_open_complaints'].unique()

array([0., 3., 2., 1., 4., 5.])

In [31]:
mkt['number_of_open_complaints'].dtype

dtype('float64')

In [32]:
mkt['number_of_policies'].unique()

array([5, 1, 2, 7, 3, 9, 8, 4, 6])

In [33]:
mkt['policy_type'].unique()

array(['Personal Auto', 'Corporate Auto', 'Special Auto'], dtype=object)

In [34]:
mkt['vehicle_class'].unique()

array(['SUV', 'Two-Door Car', 'Four-Door Car', 'Luxury SUV', 'Sports Car',
       'Luxury Car'], dtype=object)

In [35]:
mkt['vehicle_size'].unique()

array(['Medsize', 'Small', 'Large'], dtype=object)

In [37]:
mkt['policy'].unique()

array(['Personal L3', 'Personal L2', 'Corporate L2', 'Personal L1',
       'Corporate L1', 'Special L3', 'Corporate L3', 'Special L1',
       'Special L2'], dtype=object)

In [38]:
mkt['renew_offer_type'].unique()

array(['Offer1', 'Offer2', 'Offer4', 'Offer3'], dtype=object)

In [39]:
mkt['sales_channel'].unique()

array(['Call Center', 'Branch', 'Agent', 'Web'], dtype=object)

In [40]:
mkt['total_claim_amount'].unique()

array([1258.327804,  350.4     ,  511.2     , ...,   47.061469,
        491.755368, 2027.724442])

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

In [43]:
filtered_df = mkt[(mkt['total_claim_amount'] > 1000) & (mkt['response'] == 'Yes')]
filtered_df

customer       state  customer_lifetime_value response  coverage  \
127    AJ89108  California             11009.130490      Yes   Premium   
927    CG15505  California             19160.989940      Yes  Extended   
999    CK39096      Oregon              5619.689084      Yes   Premium   
1565   DT77901      Oregon             10179.717040      Yes   Premium   
2134   FD73388     Arizona             13736.132500      Yes   Premium   
2186   FG29932  California              7840.165778      Yes  Extended   
2221   FH77504  California             11009.130490      Yes   Premium   
2253   FJ28407  California              5619.689084      Yes   Premium   
2320   FN44127      Oregon              3508.569533      Yes  Extended   
2562   GA72966      Oregon              8427.172486      Yes  Extended   
2608   GD55093  California             19160.989940      Yes  Extended   
2815   GP83239  California              7840.165778      Yes  Extended   
2924   GW43195      Oregon             25807.063000      Yes  Extended   
3629   IP94270     Arizona             13736.132500      Yes   Premium   
3702   IV54766      Oregon             19160.989940      Yes  Extended   
3740   IX60941      Oregon              3508.569533      Yes  Extended   
3807   JC11405      Oregon             10963.957230      Yes   Premium   
3942   JL21509  California              3508.569533      Yes  Extended   
4243   KC26486     Arizona              8427.172486      Yes  Extended   
4963   LT72149      Nevada              5619.689084      Yes   Premium   
5083   MA15172  California             10179.717040      Yes   Premium   
5119   MC30750      Oregon             13736.132500      Yes   Premium   
5223   MJ51323      Oregon              8427.172486      Yes  Extended   
5290   MO21402      Oregon             13736.132500      Yes   Premium   
5366   MS18916  California              3508.569533      Yes  Extended   
5386   MT27589      Oregon              7840.165778      Yes  Extended   
5867   NZ60700  Washington             10963.957230      Yes   Premium   
6043   OK31456  California             11009.130490      Yes   Premium   
6499   PN98247  California              7840.165778      Yes  Extended   
6782   QE79527  California              8427.172486      Yes  Extended   
6885   QL45827  Washington             25807.063000      Yes  Extended   
6946   QO62792      Oregon              7840.165778      Yes  Extended   
7082   QW47320  California             10179.717040      Yes   Premium   
7365   RO89785      Oregon              3508.569533      Yes  Extended   
7953   TA66375      Oregon             11009.130490      Yes   Premium   
7985   TC88986      Oregon              5619.689084      Yes   Premium   
7987   TC99043  California             13736.132500      Yes   Premium   
8207   TQ84956  California             10963.957230      Yes   Premium   
8260   TU53781      Oregon              8427.172486      Yes  Extended   
8324   TY31217  Washington             11009.130490      Yes   Premium   
8461   UH35128      Oregon             25807.063000      Yes  Extended   
8603   UP29839      Nevada             19160.989940      Yes  Extended   
8669   UT48216      Nevada              7840.165778      Yes  Extended   
8680   UU42868  California              5619.689084      Yes   Premium   
8747   UY18770  California             10179.717040      Yes   Premium   
8957   VL84149      Oregon             25807.063000      Yes  Extended   
9082   VV47792  California             25807.063000      Yes  Extended   
9176   WA45944     Arizona             25807.063000      Yes  Extended   
9447   WP58340     Arizona             10963.957230      Yes   Premium   
9954   XT93203  California             10963.957230      Yes   Premium   
10034  XZ64172      Oregon             10963.957230      Yes   Premium   
10111  YE24562  California              3508.569533      Yes  Extended   
10183  YJ16163      Oregon             11009.130490      Yes   Premium   
10205  YK6

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

In [44]:
average_claims = mkt[mkt['response'] == 'Yes'].groupby(['policy_type', 'gender'])['total_claim_amount'].mean()
average_claims

policy_type     gender
Corporate Auto  F         431.480117
                M         412.756888
Personal Auto   F         454.089997
                M         453.603715
Special Auto    F         455.648867
                M         414.799291
Name: total_claim_amount, dtype: float64

The average total claim amounts vary across policy types and genders. **Personal Auto policies** have the highest claim amounts for both genders, with minimal differences between males ($453.60) and females ($454.09). **Special Auto policies** show higher claims for females ($455.65) compared to males ($414.80). On the other hand, **Corporate Auto policies** have the lowest claim amounts, especially for males ($412.76), suggesting different claim patterns in corporate settings. Overall, females tend to have slightly higher claim amounts across all policy types, except for **Personal Auto**, where the difference is negligible.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

In [47]:
customers_by_state = mkt.groupby('state')['customer'].count()
customers_by_state

state
Arizona       1703
California    3150
Nevada         882
Oregon        2601
Washington     798
Name: customer, dtype: int64

In [48]:
filtered_cbs = customers_by_state[customers_by_state > 500]
filtered_cbs

state
Arizona       1703
California    3150
Nevada         882
Oregon        2601
Washington     798
Name: customer, dtype: int64

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [49]:
result = mkt.groupby(['gender', 'education'])['customer_lifetime_value'].agg(['max', 'min', 'median'])
result

max          min       median
gender education                                                  
F      Bachelor              73225.95652  1904.000852  5678.050167
       College               61850.18803  1898.683686  5621.789489
       Doctor                44856.11397  2395.570000  5332.462694
       High School or Below  55277.44589  2144.921535  6044.020038
       Master                51016.06704  2417.777032  5801.128252
M      Bachelor              67907.27050  1898.007675  5555.830205
       College               61134.68307  1918.119700  5989.773931
       Doctor                32677.34284  2267.604038  5620.587184
       High School or Below  83325.38119  1940.981221  6176.700392
       Master                50568.25912  2272.307310  5617.956917

Females generally have a lower customer lifetime value (CLV) compared to males across most education levels. The highest CLV for females is observed in the Bachelor's degree group ($73,225.96), while for males, it is seen in the High School or Below category ($83,325.38). Minimum CLV values are relatively close for both genders, but females tend to have slightly lower minimum values. The median CLV is also similar across most education levels, with a slight variation between males and females. Overall, males tend to have higher CLV, particularly for those with lower educational attainment, while females show more consistency across education levels.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

In [63]:
mkt.head()

customer       state  customer_lifetime_value response  coverage  \
0  AA10041  California              7901.744238       No     Basic   
3  AA11235      Nevada              2568.843396       No     Basic   
5  AA16582  Washington             24127.504020      Yes     Basic   
6  AA30683  California              6595.101921       No   Premium   
7  AA34092  California             28799.954150       No  Extended   

              education effective_to_date employmentstatus gender  income  \
0  High School or Below            2/2/11       Unemployed      M       0   
3              Bachelor           1/28/11    Medical Leave      F   11167   
5              Bachelor           1/26/11    Medical Leave      M   14072   
6              Bachelor           1/15/11       Unemployed      M       0   
7               College           2/11/11         Employed      M   33635   

  location_code marital_status  monthly_premium_auto  months_since_last_claim  \
0      Suburban         Single                   113                     35.0   
3      Suburban        Married                    73                      0.0   
5      Suburban       Divorced                    71                     13.0   
6      Suburban         Single                   105                      2.0   
7      Suburban        Married                   240                      1.0   

   months_since_policy_inception  number_of_open_complaints  \
0                             83                        0.0   
3                             99                        3.0   
5                              3                        0.0   
6                             38                        2.0   
7                             23                        0.0   

   number_of_policies    policy_type       policy renew_offer_type  \
0                   5  Personal Auto  Personal L3           Offer1   
3                   1  Personal Auto  Personal L3           Offer2   
5                   2  Personal Auto  Personal L2           Offer1   
6                   7  Personal Auto  Personal L3           Offer1   
7                   2  Personal Auto  Personal L3           Offer1   

  sales_channel  total_claim_amount  vehicle_class vehicle_size  
0   Call Center         1258.327804            SUV      Medsize  
3        Branch          350.400000   Two-Door Car      Medsize  
5         Agent          511.200000  Four-Door Car      Medsize  
6           Web          847.717402  Four-Door Car      Medsize  
7           Web         1152.000000     Luxury SUV      Medsize

In [66]:
policies = mkt.pivot_table(index='state', columns='vehicle_size', values='number_of_policies', aggfunc='sum')

policies

vehicle_size  Large  Medsize  Small
state                              
Arizona         508     3665   1051
California      921     6645   1782
Nevada          186     1739    524
Oregon          805     5411   1519
Washington      249     1655    433

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

In [70]:
policies['total'] = policies['Large'] + policies['Medsize'] + policies['Small']
policies_sorted = policies.sort_values(by='total', ascending=False)
policies_sorted

vehicle_size  Large  Medsize  Small  total
state                                     
California      921     6645   1782   9348
Oregon          805     5411   1519   7735
Arizona         508     3665   1051   5224
Nevada          186     1739    524   2449
Washington      249     1655    433   2337

In [72]:
top_3_states = policies_sorted.head(3)
top_3_states

vehicle_size  Large  Medsize  Small  total
state                                     
California      921     6645   1782   9348
Oregon          805     5411   1519   7735
Arizona         508     3665   1051   5224

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

In [55]:
# Grouping by sales_channel and response, then counting the responses for each combination
mkt_channels = mkt.groupby(['sales_channel', 'response'])['response'].count().unstack(fill_value=0)

In [56]:
mkt_channels

response         No  Yes
sales_channel           
Agent          2811  666
Branch         2273  294
Call Center    1573  192
Web            1169  156

In [59]:
mkt_channels = mkt_channels.reset_index()

In [60]:
# Melting the data to unpivot it and have response values in a single column
melted_data = pd.melt(mkt_channels, id_vars=['sales_channel'], value_vars=['Yes', 'No'], 
                      var_name='Response', value_name='Count')

In [61]:
melted_data

sales_channel Response  Count
0         Agent      Yes    666
1        Branch      Yes    294
2   Call Center      Yes    192
3           Web      Yes    156
4         Agent       No   2811
5        Branch       No   2273
6   Call Center       No   1573
7           Web       No   1169

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9